src/common/scaled_scene_buffer.c

```diff
@@ -0,0 +1,177 @@
+// SPDX-License-Identifier: GPL-2.0-only
+#define _POSIX_C_SOURCE 200809L
+#include <assert.h>
+#include <stdlib.h>
+#include <wayland-server-core.h>
+#include <wlr/types/wlr_buffer.h>
+#include <wlr/types/wlr_scene.h>
+#include <wlr/util/log.h>
+#include "buffer.h"
+#include "common/scaled_scene_buffer.h"
+
+/* Internal API */
+static void
+_cache_entry_destroy(struct scaled_scene_buffer_cache_entry *cache_entry)
+{
+	wl_list_remove(&cache_entry->link);
+	if (cache_entry->buffer) {
+		wlr_buffer_drop(cache_entry->buffer);
+	}
+	free(cache_entry);
+}
+
+static void
+_update_buffer(struct scaled_scene_buffer *self, double scale)
+{
+	self->active_scale = scale;
+
+	/* Search for cached buffer of specified scale */
+	struct scaled_scene_buffer_cache_entry *cache_entry, *cache_entry_tmp;
+	wl_list_for_each_safe(cache_entry, cache_entry_tmp, &self->cache, link) {
+		if (cache_entry->scale == scale) {
+			wlr_log(WLR_ERROR, "[%p] Using cached buffer for scale %f",
+				&self->scene_buffer->node, scale);
+			/* LRU cache, recently used in front */
+			wl_list_remove(&cache_entry->link);
+			wl_list_insert(&self->cache, &cache_entry->link);
+			wlr_scene_buffer_set_buffer(self->scene_buffer, cache_entry->buffer);
+			return;
+		}
+	}
+
+	/* Limit amount of cached buffers */
+	if (wl_list_length(&self->cache) == LAB_SCALED_BUFFER_MAX_CACHE) {
+		/* Remove last element */
+		cache_entry = wl_container_of(self->cache.prev, cache_entry, link);
+		wlr_log(WLR_ERROR, "[%p] Removing oldest cache entry for scale %f",
+			&self->scene_buffer->node, cache_entry->scale);
+		_cache_entry_destroy(cache_entry);
+	}
```

Comment:
^ reuse cache entry instead of destroying. else `cache_entry = calloc(1, sizeof(*cache_entry));`
+ remove calloc of cache_entry below